src/main/java/org/opensearch/neuralsearch/util/ProcessorDocumentUtils.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.neuralsearch.util;
+
+import org.apache.commons.lang3.StringUtils;
+import org.opensearch.cluster.metadata.IndexMetadata;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.core.common.util.CollectionUtils;
+import org.opensearch.env.Environment;
+import org.opensearch.index.mapper.MapperService;
+
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * This class is used to accommodate the common code pieces of parsing, validating and processing the document for multiple
+ * pipeline processors.
+ */
+public class ProcessorDocumentUtils {
+
+    /**
+     * Validates a map type value recursively up to a specified depth. Supports Map type, List type and String type.
+     * If current sourceValue is Map or List type, recursively validates its values, otherwise validates its value.
+     *
+     * @param  sourceKey    the key of the source map being validated, the first level is always the "field_map" key.
+     * @param  sourceValue  the source map being validated, the first level is always the sourceAndMetadataMap.
+     * @param  fieldMap     the configuration map for validation, the first level is always the value of "field_map" in the processor configuration.
+     * @param  depth        the current depth of recursion
+     * @param  clusterService cluster service passed from OpenSearch core.
+     * @param  environment   environment passed from OpenSearch core.
+     * @param  indexName     the maximum allowed depth for recursion
+     * @param  allowEmpty   flag to allow empty values in map type validation.
+     */
+    @SuppressWarnings({ "rawtypes", "unchecked" })
+    public static void validateMapTypeValue(
+        final String sourceKey,
+        final Map<String, Object> sourceValue,
+        final Object fieldMap,
+        final long depth,
+        final String indexName,
+        final ClusterService clusterService,
+        final Environment environment,
```

Comment:
if clusterService and environment are only used here for those processors; how about make the 2 variables static field for ProcessorDocumentUtils? Otherwise for every new processor we need to create 2 fields for them

Reply:
It's true that we can do this to reduce code repetition, but the current repetition isn't significant, and we're adding new processors at a very slow pace. It's fine to keep things as they are now, we can optimize when more processors are added.